enhancements/kube-apiserver/kms-encryption-provider.md

```diff
@@ -0,0 +1,280 @@
+  - implements a distributed state machine to transition only takes place when all API servers have converged to the same revision
+3. Prune controller
+  - prunes inactive keys present in the cluster
+  - ensures that at any given time N keys are present (today `N=10` and hardcoded i.e. non-configurable)
+4. Migration controller
+  - mark that all the resources as migrated once they have been rewritten in etcd with desired encryption state
+5. Condition controller
+  - decides whether the other controllers should start acting upon depending upon the current &/ desired encryption mode
+  - provide for status conditions that an operator can write to their respective status field(s) eg. EncryptionInProgress, EncryptionCompleted, etc.  
+For more details about the design of the existing encryption controller(s), refer to the related enhancement: https://github.com/openshift/enhancements/blob/master/enhancements/kube-apiserver/encrypting-data-at-datastore-layer.md which explains all the transition state and key management/rotation procedures.
+
+For KMS the state_controller and key_controller will be extended to allow KMSv2 as an added type for the `EncryptionConfig.Resources.Providers.KMS` configuration. 
+
+Additional, to the encryption controller set we need to add the KMS plugin as separate static pod (or as a new container within the same kube-apiserver pod) managed via kube-apiserver-operator which will ensure to keep the gRPC unix socket active on the control-plane nodes. This can make use of livenessProbe and readinessProbes over sample encrypt and status calls sent to the plugin socket (eg. [probes](https://github.com/kubernetes-sigs/aws-encryption-provider?tab=readme-ov-file#deploy-the-aws-encryption-provider-plugin)) for health checks initially until we introduce new mechanisms to detect failures. The other operators need not run additional instances of the KMS plugin but can share the same hostPath mount from the master nodes to perform encrypt/decrypt at their end. In kube-apiserver-operator, a static pod deployment hook can alter the pod spec to add the kms plugin container only when KMS encryption type is enabled.
+
+## Implementation History
+
+1. Local AES encryption: https://github.com/openshift/enhancements/blob/master/enhancements/kube-apiserver/encrypting-data-at-datastore-layer.md
+2. Encryption controllers: https://github.com/openshift/library-go/tree/master/pkg/operator/encryption
+3. KMS PoC in OpenShift: https://github.com/openshift/cluster-kube-apiserver-operator/pull/1625
+4. Upstream kubernetes KMS encryption provider: https://github.com/kubernetes/apiserver/tree/master/pkg/storage/value/encrypt/envelope
+
+## Drawbacks
+
+1. Core apiserver operators need host access to mount and manage permissions for the directory where the kms plugin runs.
+2. Can cause problems during disaster recovery and backups in case KMS keys becomes unavailable after
```

Comment:
How can you mitigate this risk?

Reply:
This is still WIP, I'll elaborate on the risks from a high-level here, but https://issues.redhat.com/browse/API-1825 is a GA requirement which requires us to write a doc going in full depth about the risks and the mitigations that we have for them.
We will list all the failures scenarios there and explain in detail the actions that users need to take to detect and fix the issues. We will turn this document into KCS later on.

Reply:
If that's the place to track more, might be worth linking that issue in the GA requirements part of the doc, and saying here that you can read more in jira